File: part2-refactoring-gym/ts/step3-method-sep/massiveUmpire.ts
export class MassiveUmpire {
    public playGame(rawAnswer: string[], rawGuess: string[]): [number, number] {
        // 1. 입력 검증
        this.validateInput(rawAnswer, rawGuess);

        // 2. 파싱 (String -> Number)
        const answer = this.parseInput(rawAnswer);
        const guess = this.parseInput(rawGuess);

        // 3. 스트라이크 판정
        const strikeCount = this.countStrike(answer, guess);

        // 4. 볼 판정
        const ballCount = this.countBall(answer, guess);

        // 5. 결과 반환 [스트라이크, 볼]
        return [strikeCount, ballCount];
    }

    private validateInput(rawAnswer: string[], rawGuess: string[]){
        if (!rawAnswer || !rawGuess) {
            throw new Error("입력값은 null이나 undefined일 수 없습니다.");
        }
        if (rawAnswer.length !== 3 || rawGuess.length !== 3) {
            throw new Error("입력값의 길이가 3이어야 합니다.");
        }
    }

    private parseInput(raw: string[]): number[]{
        const input: number[] = [];

        for (let i = 0; i < 3; i++) {
            const parsed = Number(raw[i]);

            if (Number.isNaN(parsed)) {
                throw new Error("숫자만 입력 가능합니다.");
            }
            input.push(parsed);
        }

        return input;
    }

    private countStrike(answer: number[], guess: number[]): number{
        let strikeCount = 0;
        for (let i = 0; i < 3; i++) {
            if (answer[i] === guess[i]) {
                strikeCount++;
            }
        }
        return strikeCount;
    }

    private countBall(answer: number[], guess: number[]): number{
        let ballCount = 0;
        for (let i = 0; i < 3; i++) {
            for (let j = 0; j < 3; j++) {
                if (i !== j && answer[i] === guess[j]) {
                    ballCount++;
                }
            }
        }

        return ballCount;
    }
}

File: part2-refactoring-gym/java/step3-method-sep/src/main/java/badcode/MassiveUmpire.java
package badcode;

import java.util.List;

public class MassiveUmpire {

    public static final int DIGIT_COUNT = 3;

    public int[] playGame(List<String> rawAnswer, List<String> rawGuess) {
        validateInput(rawAnswer, rawGuess); // 레벨 2

        int[] answer = parseInput(rawAnswer);
        int[] guess = parseInput(rawGuess);

        int strikeCount = countStrike(answer, guess);
        int ballCount = countBall(answer, guess);

        // 5. 결과 반환 (0번 인덱스: 스트라이크, 1번 인덱스: 볼)
        return new int[]{strikeCount, ballCount};
    }

    private static void validateInput(List<String> rawAnswer, List<String> rawGuess) {
        // 1. 입력 검증
        if (rawAnswer == null || rawGuess == null) {
            throw new IllegalArgumentException("입력값은 null일 수 없습니다.");
        }
        if (rawAnswer.size() != DIGIT_COUNT || rawGuess.size() != DIGIT_COUNT) {
            throw new IllegalArgumentException("입력값의 길이가 3이어야 합니다.");
        }
    }

    private int[] parseInput(List<String> rawInput) {
        // 2. 파싱 (String -> Integer)
        int[] parsedInput = new int[DIGIT_COUNT];
        for (int index = 0; index < DIGIT_COUNT; index++) {
            try {
                parsedInput[index] = Integer.parseInt(rawInput.get(index));
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException("숫자만 입력 가능합니다.");
            }
        }
        return parsedInput;
    }

    private int countStrike(int[] answer, int[] guess) {
        // 3. 스트라이크 판정
        int strikeCount = 0;
        for (int index = 0; index < DIGIT_COUNT; index++) {
            if (answer[index] == guess[index]) {
                strikeCount++;
            }
        }
        return strikeCount;
    }

    private int countBall(int[] answer, int[] guess) {
        // 4. 볼 판정
        int ballCount = 0;
        for (int targetIndex = 0; targetIndex < DIGIT_COUNT; targetIndex++) {
            for (int guessIndex = 0; guessIndex < DIGIT_COUNT; guessIndex++) {
                if (targetIndex != guessIndex && answer[targetIndex] == guess[guessIndex]) {
                    ballCount++;
                }
            }
        }
        return ballCount;
    }
}
